cpp/include/graph.hpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+namespace cugraph {
+namespace experimental {
+
+typedef enum prop_type{PROP_UNDEF, PROP_FALSE, PROP_TRUE} PropType;
+
+struct GraphProperties {
+  bool directed;
+  bool weighted;
+  bool multigraph;
+  bool bipartite;
+  bool tree;
+  PropType has_negative_edges;
+  GraphProperties() : directed(false), weighted(false), multigraph(false), bipartite(false), tree(false), has_negative_edges(PROP_UNDEF){}
+};
+
+/**
+ * @Synopsis    Base class graphs, all but vertices and edges
+ *
+ * @tparam VT   Type of vertex (defaults to int)
+ * @tparam WT   Type of weight (defaults to float)
+ */
+template <typename WT>
+class GraphBase {
+public:
+  WT const *edge_data;     ///< edge weight
+
+  GraphProperties          prop;
+
+  size_t                   number_of_vertices;
```

Comment:
Better go with VT (or vertex_id or vertex_t or ...).

If we set VT to int32_t or int64_t, this can require comparing signed with unsigned, and generates a warning without an explicit type casting.